pkg/routes/login/root.go

```diff
@@ -0,0 +1,111 @@
+package login
+
+import (
+	"encoding/json"
+	"github.com/go-playground/validator/v10"
+	"github.com/gorilla/mux"
+	"github.com/naftulikay/golang-webapp/pkg/handlers"
+	"github.com/naftulikay/golang-webapp/pkg/interfaces"
+	"github.com/naftulikay/golang-webapp/pkg/response"
+	"github.com/naftulikay/golang-webapp/pkg/views"
+	"go.uber.org/zap"
+	null "gopkg.in/guregu/null.v4"
+	"net/http"
+	"time"
+)
+
+func ConfigureRoutes(app interfaces.App, r *mux.Router) {
+	r.HandleFunc("/login", handlers.App(app, LoginHandler)).Methods("POST")
+}
+
+// LoginHandler godoc
+// @Summary Login
+// @Tags auth
+// @Description Log a user in using email and password.
+// @Accept json
+// @Param login body views.LoginRequest true "Login Credentials"
+// @Produce json
+// @Success 200 {object} views.Response{data=views.LoginResponse}
+// @Failure 400 {object} views.Response
+// @Failure 403 {object} views.Response
+// @Router /login [post]
+func LoginHandler(app interfaces.App, w http.ResponseWriter, r *http.Request) {
```

Comment:
We're experiencing an issue here. `w` already has 200 written to it somewhere, so all of our `w.WriteHeader` commands fail.

Reply:
In the debugger, I can't really see where this is happening.

Reply:
**Discovery**: any time something writes to `w`, such as `json.NewEncoder(w).Encode(x)`, it automatically sets the status code to 200.

Reply:
This seems to be [answered here on StackOverflow](https://stackoverflow.com/questions/57828645/how-to-handle-superfluous-response-writeheader-call-in-order-to-return-500). 

In the [actual Go source code](https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/net/http/server.go;l=120-121), the documentation reads:

```golang
// If WriteHeader has not yet been called, Write calls
// WriteHeader(http.StatusOK) before writing the data. If the Header
// does not contain a Content-Type line, Write adds a Content-Type set
// to the result of passing the initial 512 bytes of written data to
// DetectContentType. Additionally, if the total size of all written
// data is under a few KB and there are no Flush calls, the
// Content-Length header is added automatically.
```

Apparently, that does mean that you _must_ call `w.WriteHeader` before any calls to `w.Write` occur.